### PR TITLE
[MC-36] Added first version of `add_app_to_policy` action.

### PR DESCRIPTION
**Pull request by tfranckiewicz-r7**
### Description

Describe the proposed changes:

  -

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

<details>

```
{
  "success": true
}
```

<summary>
docker run --rm -i rapid7/microsoft_intune:1.4.0 run < tests/add_app_to_policy.json
</summary>
</details>

<details>

```
{
  "success": true
}
```

<summary>
docker run --rm -i rapid7/microsoft_intune:1.4.0 --debug run < tests/delete_app_from_policy.json
</summary>
</details>

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Helpers.mk ../../tools/Makefiles/Colors.mk
--
make: Warning: File 'Makefile' has modification time 4007 s in the future
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
318
WARNING: Unsupported keywords found: intune, cloud. The following keywords will not be searchable by the Extension Library. Please remove or update the invalid keywords from the keywords array in the plugin.spec.yaml file.
[*] Plugin successfully validated!

----
[*] Total time elapsed: 2263.0699999999997ms
make: warning:  Clock skew detected.  Your build may be incomplete.
```

<summary>
make validate
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
318
WARNING: Unsupported keywords found: intune, cloud. The following keywords will not be searchable by the Extension Library. Please remove or update the invalid keywords from the keywords array in the plugin.spec.yaml file.
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 14217.742ms
```

<summary>
icon-validate --all . 
</summary>
</details>

### CODE REVIEW

File: microsoft_intune/help.md [hunk 1]
```diff
@@ -56,6 +56,39 @@ Example input:
 
 ### Actions
 
+#### Add application to a policy
+
+This action is used to add application to a policy.
```
Comment (jschipp-r7) [inline microsoft_intune/help.md:61]:
```suggestion
This action is used to add an application to a policy.
```

File: microsoft_intune/plugin.spec.yaml [hunk 2]
```diff
@@ -614,4 +614,40 @@ actions:
...
+        description: Application name
+        type: string
+        required: true
+        example: Adobe Reader
+      device_name:
+        title: Device
+        description: Device name
+        type: string
+        required: true
+        example:
```
Comment (jschipp-r7) [inline microsoft_intune/plugin.spec.yaml:633]: Need an example device name

File: microsoft_intune/help.md [hunk 3]
```diff
@@ -56,6 +56,39 @@ Example input:
...
+
+##### Output
+
+|Name|Type|Required|Description|
+|----|----|--------|-----------|
+|success|object|False|Return true if action was successfully performed on device|
+
+Example output:
+
+```
```
Comment (jschipp-r7) [inline microsoft_intune/help.md:89]: Need JSON formatted example output from a real request. Sanitize any data that shouldn't be viewable publicly

File: microsoft_intune/plugin.spec.yaml [hunk 4]
```diff
@@ -614,4 +614,40 @@ actions:
         title: Success
         description: Return true if action was successfully performed on device
         type: boolean
-        required: true
\ No newline at end of file
+        required: true
+  add_app_to_policy:
+    title: Add application to a policy
```
Comment (jschipp-r7) [inline microsoft_intune/plugin.spec.yaml:619]: Title style this, will need to regen after - and make sure it's reflected in help.md

```suggestion
    title: Add Application to a Policy
```

File: microsoft_intune/help.md [hunk 5]
```diff
@@ -56,6 +56,39 @@ Example input:
...
+|----|----|-------|--------|-----------|----|-------|
+|application_name|string|None|True|Application name|None|Adobe Reader|
+|device_name|string|None|True|Device name|None|None|
+|device_type|string|None|True|Device type|['Android', 'iOS']|None|
+|policy_name|string|None|True|Policy name|None|my-android-app-policy|
+
+Example input:
+
+```
+{
```
Comment (jschipp-r7) [inline microsoft_intune/help.md:75]: if all the inputs are `required: true` we should add the other two options in this JSON example

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]
@tfranckiewicz-r7

**Comment by jschipp-r7** [comment]
@tfranckiewicz-r7 there's a few merge conflicts here to fix

**Review by jmcadams-r7 — APPROVED** [review]

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Review by jmcadams-r7 — CHANGES_REQUESTED** [review]
Validation failed, Need to regen the plugin and re-commit.

File: plugins/microsoft_intune/icon_microsoft_intune/actions/delete_app_from_policy/action.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+                params.get(Input.APPLICATION_NAME), params.get(Input.POLICY_NAME), params.get(Input.DEVICE_TYPE)
+            )
+
+            if not response:
+                return {Output.SUCCESS: True}
+
+            return {Output.SUCCESS: False}
+
+        except PluginException:
+            return {Output.SUCCESS: False}
```
Comment (mberezin-r7) [inline plugins/microsoft_intune/icon_microsoft_intune/actions/delete_app_from_policy/action.py:29]: Do we want to add logging to this exception?

File: plugins/microsoft_intune/icon_microsoft_intune/util/api.py [hunk 7]
```diff
@@ -16,6 +16,96 @@ def __init__(self, username, password, client_id, client_secret, tenant_id, api_
...
+
+        if not application_package_id:
+            raise PluginException(
+                cause=f"Application: {application_name}, was not found.",
+                assistance="Contact support for help. See log for more details.",
+            )
+
+        target_apps = []
+
+        for item in managed_app_policies_with_apps["apps"]:
```
Comment (mberezin-r7) [inline plugins/microsoft_intune/icon_microsoft_intune/util/api.py:43]:
```suggestion
        for item in managed_app_policies_with_apps.get("apps"):
```
Prevents `KeyError` if "apps" not in `managed_app_policies_with_apps`

File: plugins/microsoft_intune/icon_microsoft_intune/util/api.py [hunk 8]
```diff
@@ -16,6 +16,96 @@ def __init__(self, username, password, client_id, client_secret, tenant_id, api_
...
```
Comment (mberezin-r7) [inline plugins/microsoft_intune/icon_microsoft_intune/util/api.py:92]:
```suggestion
        for item in managed_app_policies_with_apps.get("apps"):
```
Prevents `KeyError` if "apps" not in `managed_app_policies_with_apps`

File: plugins/microsoft_intune/icon_microsoft_intune/util/api.py [hunk 9]
```diff
@@ -16,6 +16,96 @@ def __init__(self, username, password, client_id, client_secret, tenant_id, api_
...
+            if item["mobileAppIdentifier"]["packageId"] != application_package_id:
```
Comment (mberezin-r7) [inline plugins/microsoft_intune/icon_microsoft_intune/util/api.py:93]: We should verify that these keys exist in this structure. Where can we find documentation about this?

File: plugins/microsoft_intune/icon_microsoft_intune/util/api.py [hunk 10]
```diff
@@ -214,3 +304,15 @@ def create_necessary_headers(access_token):
             "Authorization": f"Bearer {access_token}",
             "Content-Type": "application/json",
         }
+
+    def _filter_policy_id(self, managed_app_policies, policy_name):
+        for item in managed_app_policies["value"]:
```
Comment (mberezin-r7) [inline plugins/microsoft_intune/icon_microsoft_intune/util/api.py:309]:
```suggestion
        for item in managed_app_policies.get("value"):
```
Prevent `KeyError` if `"value"` not in `managed_app_policies`

File: plugins/microsoft_intune/icon_microsoft_intune/util/api.py [hunk 11]
```diff
@@ -214,3 +304,15 @@ def create_necessary_headers(access_token):
...
+            if item["displayName"] == policy_name:
+                return item["id"]
+        return None
+
+    def _filter_app_package_id(self, managed_app_list, application_name):
+        for item in managed_app_list["content"]["appList"]:
```
Comment (mberezin-r7) [inline plugins/microsoft_intune/icon_microsoft_intune/util/api.py:315]: We should also verify that this structure is consistent and we can hardcode these as constants. In which case, we should use programatic constants.

**Pull request closed without merging**